{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f0a72014-5773-4b26-8a57-a13839dd0654",
   "metadata": {},
   "source": [
    "Source of original data: [https://genomics.senescence.info](https://genomics.senescence.info/longevity/)<br />\n",
    "*(LongevityMap build 3, release date: 2017 June 24, number of genes: 884)*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4b20f4a6-3c4b-4e27-9dd9-bb8636702473",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "94d323f6-6cd9-4a01-a8b1-5c908585918e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (550, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-B40</td>\n",
       "      <td>HLA-B</td>\n",
       "      <td>1859103</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>1859103</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOB</td>\n",
       "      <td>APOB</td>\n",
       "      <td>8018664</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>8018664</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>E2/E3/E4</td>\n",
       "      <td>APOE</td>\n",
       "      <td>8018664</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id      Association Population Variant(s)   Gene(s)   PubMed  Unnamed: 6\n",
       "0  1  non-significant      Dutch    HLA-B40     HLA-B  1859103         NaN\n",
       "1  2  non-significant      Dutch   HLA-DRB5  HLA-DRB5  1859103         NaN\n",
       "2  3  non-significant    Finnish       APOB      APOB  8018664         NaN\n",
       "3  4      significant    Finnish      APOC3     APOC3  8018664         NaN\n",
       "4  5      significant    Finnish   E2/E3/E4      APOE  8018664         NaN"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read original data into dataFrame\n",
    "df = pd.read_csv('..\\data\\input\\longevity_genes_original.csv')\n",
    "\n",
    "print(\"Dimension:\", df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bca2ec0-3fae-42ee-891a-55a4be9bc8c0",
   "metadata": {},
   "source": [
    "<br />\n",
    "\n",
    "### Clean some data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "612381d8-930a-4f0a-bc94-8142afc23718",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In the explored column, number of records with comma at the end: 1  - cleaned\n"
     ]
    }
   ],
   "source": [
    "# if string in the column 'Variant(s)' contains comma at the end, delete this last character (ignore records that contains NaN)\n",
    "print(\"In the explored column, number of records with comma at the end:\",\n",
    "      len(df[df['Variant(s)'].str.endswith(',', na=False)]), end='')\n",
    "\n",
    "df['Variant(s)'] = df['Variant(s)'].map(lambda st: st[:-1] if st.endswith(',') else st, na_action='ignore')\n",
    "\n",
    "# check that we eliminate all cases with comma at the end\n",
    "assert len(df[df['Variant(s)'].str.endswith(',', na=False)]) == 0, \"Something wrong with removing comma at the end of string\"\n",
    "print('  - cleaned')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6c743b86-c4ea-4c29-8787-de34d52927e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In the explored column, number of records containing two commas in a row: 2  - cleaned\n"
     ]
    }
   ],
   "source": [
    "# if string in the column 'Variant(s)' contains two commas in a row, replace them by one comma\n",
    "print(\"In the explored column, number of records containing two commas in a row:\",\n",
    "      len(df[df['Variant(s)'].str.contains(',,', na=False)]), end='')\n",
    "\n",
    "df['Variant(s)'] = df['Variant(s)'].str.replace(',,', ',')\n",
    "\n",
    "# check that we eliminate all cases with two commas in a row\n",
    "assert len(df[df['Variant(s)'].str.contains(',,', na=False)]) == 0, \"Something wrong with removing two commas in a row\"\n",
    "print('  - cleaned')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5b57ef5-dea4-4833-8f23-5ac60068275c",
   "metadata": {},
   "source": [
    "<br />\n",
    "\n",
    "### Preliminary separation of original dataFrame into several dataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6aa857bc-5345-4a12-9bc2-db500ea3994e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (550, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-B40</td>\n",
       "      <td>HLA-B</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOB</td>\n",
       "      <td>APOB</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>E2/E3/E4</td>\n",
       "      <td>APOE</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id      Association Population Variant(s)   Gene(s)   PubMed\n",
       "0  1  non-significant      Dutch    HLA-B40     HLA-B  1859103\n",
       "1  2  non-significant      Dutch   HLA-DRB5  HLA-DRB5  1859103\n",
       "2  3  non-significant    Finnish       APOB      APOB  8018664\n",
       "3  4      significant    Finnish      APOC3     APOC3  8018664\n",
       "4  5      significant    Finnish   E2/E3/E4      APOE  8018664"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Last columns of the dataFrame contains only NaN value.\n",
    "# Make sure of this and delete this column as unnecessary.\n",
    "assert df['Unnamed: 6'].isna().all(), \"In the column 'Unnamed: 6' there is at least 1 record with value other than Nan\"\n",
    "del df['Unnamed: 6']\n",
    "\n",
    "print(\"Dimension:\", df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "000ed94e-df60-4d50-abe2-c7c0f181cdcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (16, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>258</td>\n",
       "      <td>significant</td>\n",
       "      <td>American (Caucasian)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>20824210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>261</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>American (Caucasian)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>20824210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>280</td>\n",
       "      <td>significant</td>\n",
       "      <td>European</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>23286790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>281</td>\n",
       "      <td>significant</td>\n",
       "      <td>European</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>23286790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>282</td>\n",
       "      <td>significant</td>\n",
       "      <td>European</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>23286790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id      Association            Population Variant(s) Gene(s)    PubMed\n",
       "255  258      significant  American (Caucasian)        NaN    TP53  20824210\n",
       "256  261  non-significant  American (Caucasian)        NaN    TP53  20824210\n",
       "275  280      significant              European        NaN    TP53  23286790\n",
       "276  281      significant              European        NaN    TP53  23286790\n",
       "277  282      significant              European        NaN    TP53  23286790"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Seperate records that do not have values in the column 'Variant(s)'\n",
    "df_var_nan = df[(df['Variant(s)'].isna())]\n",
    "\n",
    "print(\"Dimension:\", df_var_nan.shape)\n",
    "df_var_nan.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3e787cf1-d01d-4844-8ec5-6df5742d3b08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (534, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-B40</td>\n",
       "      <td>HLA-B</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOB</td>\n",
       "      <td>APOB</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>E2/E3/E4</td>\n",
       "      <td>APOE</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id      Association Population Variant(s)   Gene(s)   PubMed\n",
       "0  1  non-significant      Dutch    HLA-B40     HLA-B  1859103\n",
       "1  2  non-significant      Dutch   HLA-DRB5  HLA-DRB5  1859103\n",
       "2  3  non-significant    Finnish       APOB      APOB  8018664\n",
       "3  4      significant    Finnish      APOC3     APOC3  8018664\n",
       "4  5      significant    Finnish   E2/E3/E4      APOE  8018664"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rows that have some info in the column 'Variant(s)' - this is intermediate dataFrame for further splitting\n",
    "df_var_not_nan = df[(df['Variant(s)'].notna())]\n",
    "\n",
    "print(\"Dimension:\", df_var_not_nan.shape)\n",
    "df_var_not_nan.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7130cfb1-fc28-4fe0-bd67-85b5b474d6dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (320, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-B40</td>\n",
       "      <td>HLA-B</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOB</td>\n",
       "      <td>APOB</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>E2/E3/E4</td>\n",
       "      <td>APOE</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id      Association Population Variant(s)   Gene(s)   PubMed\n",
       "0  1  non-significant      Dutch    HLA-B40     HLA-B  1859103\n",
       "1  2  non-significant      Dutch   HLA-DRB5  HLA-DRB5  1859103\n",
       "2  3  non-significant    Finnish       APOB      APOB  8018664\n",
       "3  4      significant    Finnish      APOC3     APOC3  8018664\n",
       "4  5      significant    Finnish   E2/E3/E4      APOE  8018664"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rows that have sting in the column 'Variant(s)'\n",
    "# but this string does not contains characters 'rs' with two digits after them\n",
    "df_var_not_rs = df_var_not_nan[~df_var_not_nan['Variant(s)'].str.contains('.*rs\\d{2}.*', flags=re.IGNORECASE)]\n",
    "\n",
    "print(\"Dimension:\", df_var_not_rs.shape)\n",
    "df_var_not_rs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e83d7343-ef20-4cb3-a23b-6055e240412a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (214, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>G548</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>rs1685354,rs647126</td>\n",
       "      <td>UCP3</td>\n",
       "      <td>22743239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>G549</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Jordanian</td>\n",
       "      <td>rs2241766,rs266729</td>\n",
       "      <td>ADIPOQ</td>\n",
       "      <td>20201642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs6457931,rs1321312,rs4331968,rs9470367,rs6920...</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>G552</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>rs2866164,Q95H</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>16015282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>546</th>\n",
       "      <td>556</td>\n",
       "      <td>significant</td>\n",
       "      <td>American (Caucasian)</td>\n",
       "      <td>rs1042714</td>\n",
       "      <td>ADRB2</td>\n",
       "      <td>20399803</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id      Association            Population  \\\n",
       "539  G548  non-significant                Danish   \n",
       "540  G549  non-significant             Jordanian   \n",
       "541  G550  non-significant               Italian   \n",
       "543  G552  non-significant                Danish   \n",
       "546   556      significant  American (Caucasian)   \n",
       "\n",
       "                                            Variant(s)              Gene(s)  \\\n",
       "539                                 rs1685354,rs647126                 UCP3   \n",
       "540                                 rs2241766,rs266729               ADIPOQ   \n",
       "541  rs6457931,rs1321312,rs4331968,rs9470367,rs6920...  PANDAR,CDKN1A,RAB44   \n",
       "543                                     rs2866164,Q95H                 MTTP   \n",
       "546                                          rs1042714                ADRB2   \n",
       "\n",
       "       PubMed  \n",
       "539  22743239  \n",
       "540  20201642  \n",
       "541  20126416  \n",
       "543  16015282  \n",
       "546  20399803  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rows that have sting in the column 'Variant(s)' that contains characters 'rs' with two digits after them\n",
    "df_var_rs = df_var_not_nan[df_var_not_nan['Variant(s)'].str.contains('.*rs\\d{2}.*', flags=re.IGNORECASE)]\n",
    "\n",
    "print(\"Dimension:\", df_var_rs.shape)\n",
    "df_var_rs.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a7402eb9-d2ec-4395-a0f9-3875ef1daf29",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ensure that we did not lost something during splitting of the original dataFrame\n",
    "assert len(df) == len(df_var_nan) + len(df_var_rs) + len(df_var_not_rs), \"Something wrong with splitting\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e95c29df-2e00-483b-9971-16072d49f6b2",
   "metadata": {},
   "source": [
    "<br />\n",
    "\n",
    "### Splitting of records with several 'rs' into several records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d810269b-c640-411c-b641-d7e407588d6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of list: 214\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[2, 2, 31, 2, 1]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of variants in each record in dataFrame with rs-variants.\n",
    "# It is calculated by the number of commas +1\n",
    "nmb_repeats = (df_var_rs['Variant(s)'].str.count(',') + 1).tolist()\n",
    "\n",
    "assert len(df_var_rs) == len(nmb_repeats), \"Problem with calculation\"\n",
    "print('Length of list:', len(nmb_repeats))\n",
    "nmb_repeats[-5:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0b766727-07c6-4690-b4ac-a89af5be1693",
   "metadata": {},
   "outputs": [],
   "source": [
    "## alternative way:\n",
    "# nmb_repeats_alt = [len(st.split(',')) for st in df_var_rs['Variant(s)']]\n",
    "\n",
    "## make sure that two ways gives the same result\n",
    "# assert len(nmb_repeats_alt) == len(nmb_repeats), \"Problems with calculation\"\n",
    "# assert nmb_repeats_alt == nmb_repeats, \"Problems with calculation (2)\"\n",
    "# nmb_repeats_alt[-5:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "29a661aa-5ca5-4898-81ef-01d87a8772a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (3032, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs6457931,rs1321312,rs4331968,rs9470367,rs6920...</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs6457931,rs1321312,rs4331968,rs9470367,rs6920...</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>G552</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>rs2866164,Q95H</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>16015282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>G552</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>rs2866164,Q95H</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>16015282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>546</th>\n",
       "      <td>556</td>\n",
       "      <td>significant</td>\n",
       "      <td>American (Caucasian)</td>\n",
       "      <td>rs1042714</td>\n",
       "      <td>ADRB2</td>\n",
       "      <td>20399803</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id      Association            Population  \\\n",
       "541  G550  non-significant               Italian   \n",
       "541  G550  non-significant               Italian   \n",
       "543  G552  non-significant                Danish   \n",
       "543  G552  non-significant                Danish   \n",
       "546   556      significant  American (Caucasian)   \n",
       "\n",
       "                                            Variant(s)              Gene(s)  \\\n",
       "541  rs6457931,rs1321312,rs4331968,rs9470367,rs6920...  PANDAR,CDKN1A,RAB44   \n",
       "541  rs6457931,rs1321312,rs4331968,rs9470367,rs6920...  PANDAR,CDKN1A,RAB44   \n",
       "543                                     rs2866164,Q95H                 MTTP   \n",
       "543                                     rs2866164,Q95H                 MTTP   \n",
       "546                                          rs1042714                ADRB2   \n",
       "\n",
       "       PubMed  \n",
       "541  20126416  \n",
       "541  20126416  \n",
       "543  16015282  \n",
       "543  16015282  \n",
       "546  20399803  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dublicate records in the dataFrame 'df_var_rs', that have several variants in the column 'Variant(s)',\n",
    "# according to the number of these variants.\n",
    "df_var_rs_extended = df_var_rs.loc[df_var_rs.index.repeat(nmb_repeats)]\n",
    "\n",
    "print(\"Dimension:\", df_var_rs_extended.shape)\n",
    "df_var_rs_extended.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c72fdf63-a025-4358-a92b-415f00d54b95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<generator object <genexpr> at 0x00000226196C4A50>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create \"list\" of all variants in the column 'Variant(s)' in the dataFrame 'df_var_rs'.\n",
    "# To speed up calculation, use generator.\n",
    "gen_variants = (el for ls in df_var_rs['Variant(s)'] for el in ls.split(','))\n",
    "gen_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "97e8a8fe-ac1d-414e-82f8-9e49f39147bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Look at result (warning: the generator will be exhausted)\n",
    "# ls_variants = list(gen_variants)\n",
    "# print('Length of resulting list:', len(ls_variants))\n",
    "# ls_variants[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a59b96b8-28c0-4643-aeb6-eb500f01cd6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## alternative way:\n",
    "# import itertools\n",
    "# iter_variants = itertools.chain.from_iterable(ls.split(',') for ls in df_var_rs['Variant(s)'])\n",
    "# print(type(iter_variants))\n",
    "\n",
    "## look at result (warning: the iterator will be exhausted) and make sure that two ways gives the same result\n",
    "# ls_variants_alt = list(iter_variants)\n",
    "\n",
    "# assert len(ls_variants_alt) == len(ls_variants), \"Problem with calculation\"\n",
    "# print('Length of resulting list:', len(ls_variants_alt))\n",
    "\n",
    "# assert ls_variants_alt == ls_variants, \"Problem with calculation (2)\"\n",
    "# ls_variants_alt[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2a493e1d-e4f6-4b3d-9dba-8a0ce3585706",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (3032, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs6457940</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs2145047</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>G552</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>rs2866164</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>16015282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>G552</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>Q95H</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>16015282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>546</th>\n",
       "      <td>556</td>\n",
       "      <td>significant</td>\n",
       "      <td>American (Caucasian)</td>\n",
       "      <td>rs1042714</td>\n",
       "      <td>ADRB2</td>\n",
       "      <td>20399803</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id      Association            Population Variant(s)  \\\n",
       "541  G550  non-significant               Italian  rs6457940   \n",
       "541  G550  non-significant               Italian  rs2145047   \n",
       "543  G552  non-significant                Danish  rs2866164   \n",
       "543  G552  non-significant                Danish       Q95H   \n",
       "546   556      significant  American (Caucasian)  rs1042714   \n",
       "\n",
       "                 Gene(s)    PubMed  \n",
       "541  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541  PANDAR,CDKN1A,RAB44  20126416  \n",
       "543                 MTTP  16015282  \n",
       "543                 MTTP  16015282  \n",
       "546                ADRB2  20399803  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# change content of the columns 'Variant(s)' by splitting data\n",
    "df_var_rs_extended['Variant(s)'] = list(gen_variants)\n",
    "\n",
    "print(\"Dimension:\", df_var_rs_extended.shape)\n",
    "df_var_rs_extended.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6567523c-7a32-45df-a0a1-27790d64e27d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# in the column 'Variant(s)' of dataFrame 'df_var_rs_extended', if the first two characters are 'Rs' than change them to 'rs'\n",
    "df_var_rs_extended['Variant(s)'] = df_var_rs_extended['Variant(s)'].map(lambda st: 'rs' + st[2:] if st.startswith('Rs') else st)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0e826ea-a7ec-4e6c-a7bb-6bfa55e00325",
   "metadata": {},
   "source": [
    "<br />\n",
    "\n",
    "### Combining dataFrames and saving result to disk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4da7e5a0-02aa-4654-b7d0-1bf3dae8959a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (3368, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs3176343</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs3176344</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs3176349</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs876581</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs6457938</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs6457940</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs2145047</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>G552</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>rs2866164</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>16015282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>G552</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>Q95H</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>16015282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>546</th>\n",
       "      <td>556</td>\n",
       "      <td>significant</td>\n",
       "      <td>American (Caucasian)</td>\n",
       "      <td>rs1042714</td>\n",
       "      <td>ADRB2</td>\n",
       "      <td>20399803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-B40</td>\n",
       "      <td>HLA-B</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOB</td>\n",
       "      <td>APOB</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>E2/E3/E4</td>\n",
       "      <td>APOE</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id      Association            Population Variant(s)  \\\n",
       "541  G550  non-significant               Italian  rs3176343   \n",
       "541  G550  non-significant               Italian  rs3176344   \n",
       "541  G550  non-significant               Italian  rs3176349   \n",
       "541  G550  non-significant               Italian   rs876581   \n",
       "541  G550  non-significant               Italian  rs6457938   \n",
       "541  G550  non-significant               Italian  rs6457940   \n",
       "541  G550  non-significant               Italian  rs2145047   \n",
       "543  G552  non-significant                Danish  rs2866164   \n",
       "543  G552  non-significant                Danish       Q95H   \n",
       "546   556      significant  American (Caucasian)  rs1042714   \n",
       "0       1  non-significant                 Dutch    HLA-B40   \n",
       "1       2  non-significant                 Dutch   HLA-DRB5   \n",
       "2       3  non-significant               Finnish       APOB   \n",
       "3       4      significant               Finnish      APOC3   \n",
       "4       5      significant               Finnish   E2/E3/E4   \n",
       "\n",
       "                 Gene(s)    PubMed  \n",
       "541  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541  PANDAR,CDKN1A,RAB44  20126416  \n",
       "543                 MTTP  16015282  \n",
       "543                 MTTP  16015282  \n",
       "546                ADRB2  20399803  \n",
       "0                  HLA-B   1859103  \n",
       "1               HLA-DRB5   1859103  \n",
       "2                   APOB   8018664  \n",
       "3                  APOC3   8018664  \n",
       "4                   APOE   8018664  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combine all dataFrames in one\n",
    "df_result_all = pd.concat([df_var_rs_extended, df_var_not_rs, df_var_nan])\n",
    "\n",
    "print(\"Dimension:\", df_result_all.shape)\n",
    "df_result_all.iloc[len(df_var_rs_extended)-10:len(df_var_rs_extended)+5]   # look at the middle of dataframe, where joint occured"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d0a61b18-d4cb-45b4-876b-50275cc30871",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save all results together\n",
    "df_result_all.to_csv('..\\data\\output\\longevity_genes_splitted_all.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37b1aab1-1f52-4d4e-ae67-a81446b851ff",
   "metadata": {},
   "source": [
    "<br />"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f5816cc1-1a52-4e7a-99b9-77d4e30a81b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (3025, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs6457938</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs6457940</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs2145047</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>G552</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>rs2866164</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>16015282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>546</th>\n",
       "      <td>556</td>\n",
       "      <td>significant</td>\n",
       "      <td>American (Caucasian)</td>\n",
       "      <td>rs1042714</td>\n",
       "      <td>ADRB2</td>\n",
       "      <td>20399803</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id      Association            Population Variant(s)  \\\n",
       "541  G550  non-significant               Italian  rs6457938   \n",
       "541  G550  non-significant               Italian  rs6457940   \n",
       "541  G550  non-significant               Italian  rs2145047   \n",
       "543  G552  non-significant                Danish  rs2866164   \n",
       "546   556      significant  American (Caucasian)  rs1042714   \n",
       "\n",
       "                 Gene(s)    PubMed  \n",
       "541  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541  PANDAR,CDKN1A,RAB44  20126416  \n",
       "543                 MTTP  16015282  \n",
       "546                ADRB2  20399803  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# save only records that contain 'rs' at the beginning\n",
    "df_result_rs = df_var_rs_extended[df_var_rs_extended['Variant(s)'].str.startswith('rs')]\n",
    "\n",
    "print(\"Dimension:\", df_result_rs.shape)\n",
    "df_result_rs.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "02ed50f0-f7fa-4b06-8135-d1d168ecf02f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_result_rs.to_csv('..\\data\\output\\longevity_genes_splitted_rs.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a25036d6-b6a5-4faf-82a9-0d68877dd38f",
   "metadata": {},
   "source": [
    "<br />"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "fcf0ef4c-8657-4b52-8de7-d0b60d0643ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (343, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>G414</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>German</td>\n",
       "      <td>Q/H 95</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>15911777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>415</th>\n",
       "      <td>G423</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Ashkenazi Jewish</td>\n",
       "      <td>APM1+2019</td>\n",
       "      <td>ADIPOQ</td>\n",
       "      <td>18511746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>415</th>\n",
       "      <td>G423</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Ashkenazi Jewish</td>\n",
       "      <td>APM+2019</td>\n",
       "      <td>ADIPOQ</td>\n",
       "      <td>18511746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508</th>\n",
       "      <td>G517</td>\n",
       "      <td>significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>-308G/A</td>\n",
       "      <td>TNF</td>\n",
       "      <td>18511747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>516</th>\n",
       "      <td>G525</td>\n",
       "      <td>Non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>-373(A)n(T)m</td>\n",
       "      <td>IL6</td>\n",
       "      <td>15130757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>375</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>12483296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>379</td>\n",
       "      <td>significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>10463944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>387</td>\n",
       "      <td>significant</td>\n",
       "      <td>European (Danish, Finnish, South Italian and G...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>24341918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>547</th>\n",
       "      <td>557</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>American (Caucasian)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>20824210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548</th>\n",
       "      <td>558</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>American (Caucasian)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>20824210</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>343 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       id      Association                                         Population  \\\n",
       "407  G414  non-significant                                            German    \n",
       "415  G423  non-significant                                  Ashkenazi Jewish    \n",
       "415  G423  non-significant                                  Ashkenazi Jewish    \n",
       "508  G517      significant                                            Italian   \n",
       "516  G525  Non-significant                                             Danish   \n",
       "..    ...              ...                                                ...   \n",
       "369   375      significant                                            Finnish   \n",
       "373   379      significant                                            Italian   \n",
       "381   387      significant  European (Danish, Finnish, South Italian and G...   \n",
       "547   557  non-significant                               American (Caucasian)   \n",
       "548   558  non-significant                               American (Caucasian)   \n",
       "\n",
       "       Variant(s) Gene(s)    PubMed  \n",
       "407        Q/H 95    MTTP  15911777  \n",
       "415     APM1+2019  ADIPOQ  18511746  \n",
       "415      APM+2019  ADIPOQ  18511746  \n",
       "508       -308G/A     TNF  18511747  \n",
       "516  -373(A)n(T)m     IL6  15130757  \n",
       "..            ...     ...       ...  \n",
       "369           NaN    TP53  12483296  \n",
       "373           NaN    TP53  10463944  \n",
       "381           NaN    TP53  24341918  \n",
       "547           NaN    TP53  20824210  \n",
       "548           NaN    TP53  20824210  \n",
       "\n",
       "[343 rows x 6 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# save other records, that do not contain 'rs' at the beginning\n",
    "df_result_not_rs = pd.concat([df_var_rs_extended[~df_var_rs_extended['Variant(s)'].str.startswith('rs')],\n",
    "                              df_var_not_rs,\n",
    "                              df_var_nan\n",
    "                             ])\n",
    "\n",
    "print(\"Dimension:\", df_result_not_rs.shape)\n",
    "df_result_not_rs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "97f4f76b-d8d4-467a-a62b-40a65798e054",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_result_not_rs.to_csv('..\\data\\output\\longevity_genes_splitted_not_rs.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "316a3998-03c9-4036-bb48-810deb2d1cbb",
   "metadata": {},
   "source": [
    "<br />"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "739b7f5c-5b44-4901-9903-debe2baa04b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n"
     ]
    }
   ],
   "source": [
    "# ensure that we did lost something\n",
    "assert len(df_var_rs_extended) + len(df_var_not_rs) + len(df_var_nan) == len(df_result_all), \"Something was lost\"\n",
    "assert len(df_result_all) == len(df_result_rs) + len(df_result_not_rs), \"Something was lost (2)\"\n",
    "print('Done')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
